File: apps/ios/BitOS/Platform/Feed/HashtagFollowsStore.swift
import BusinessCore
import Foundation
import Observation

/**
 * NIP-51 followed hashtags (kind 30015 d=interest) — web `hashtag-follows`
 * parity: the account's interest-set head syncs from relays (newest
 * verified list wins); toggles flip optimistically and the caller publishes
 * the resulting set through the note publisher.
 */
@MainActor
@Observable
final class HashtagFollowsStore {
    private(set) var hashtags: Set<String> = []

    private let pool: RelayPool
    private let bridge = BusinessCoreBridge()
    private var accountPubkey: String?
    private var headAt: Int64 = Int64.min
    private var framesTask: Task<Void, Never>?
    private var requested = false

    init(pool: RelayPool) {
        self.pool = pool
    }

    /// Sendable mirror of [accountPubkey]: the off-main ingest reads the
    /// live account without touching main-actor state (the bridge verifies
    /// signature, kind, d coordinate and author inside the extraction).
    private final class AccountBox: @unchecked Sendable {
        private let lock = NSLock()
        private var value: String?
        func set(_ pubkey: String?) { lock.lock(); value = pubkey; lock.unlock() }
        func get() -> String? { lock.lock(); defer { lock.unlock() }; return value }
    }

    private let accountBox = AccountBox()

    /// Newest-wins payload extracted off the main actor.
    private struct InterestHead: Sendable {
        let hashtags: [String]
        let createdAt: Int64
    }

    func start() {
        guard framesTask == nil else { return }
        let boxedBridge = StatelessBridge(bridge: bridge)
        Task { [weak self, pool] in
            let stream = await pool.verifiedFrames(client: FrameworkBusinessCoreClient())
            guard let self, !Task.isCancelled else { return }
            self.framesTask = FrameIngest.pump(
                gated: stream,
                isAlive: { [weak self] in self != nil },
                ingest: Self.headIngest(boxedBridge, accountBox: self.accountBox)
            ) { [weak self] head in
                await self?.absorbHead(head)
            }
        }
    }

    /// Account lifecycle: re-opens the interest-set head REQ.
    func setAccount(_ pubkey: String?) {
        accountPubkey = pubkey
        accountBox.set(pubkey)
        headAt = Int64.min
        requested = false
        hashtags = []
        subscribe()
    }

    /// Optimistic local flip; the caller publishes the returned set.
    func toggle(_ tag: String) -> [String] {
        let normalized = tag.trimmingCharacters(in: .whitespacesAndNewlines)
            .replacingOccurrences(of: "#", with: "")
            .lowercased()
        guard normalized.count >= 2, normalized.count <= 60 else { return Array(hashtags) }
        if hashtags.contains(normalized) {
            hashtags.remove(normalized)
        } else {
            hashtags.insert(normalized)
        }
        return Array(hashtags)
    }

    func isFollowed(_ tag: String) -> Bool {
        hashtags.contains(tag.replacingOccurrences(of: "#", with: "").lowercased())
    }

    private func subscribe() {
        guard let accountPubkey, !requested else { return }
        requested = true
        Task {
            await pool.start()
            start()
            if let request = (bridge.interestSetRequest(
                subscriptionId: "bitos-interest",
                accountPubkey: accountPubkey ?? ""
            ) as String?) {
                await pool.broadcast(request)
            }
        }
    }

    /// Interest-set head projection from an ALREADY-VERIFIED event — runs
    /// OFF the main actor via the single event-based seam (Phase 2); the
    /// bridge checks kind and author inside the call.
    private nonisolated static func headIngest(
        _ boxedBridge: StatelessBridge,
        accountBox: AccountBox
    ) -> @Sendable (GatedFrame) -> InterestHead? {
        { gated in
            guard case .event(let gatedEvent) = gated,
                  let account = accountBox.get(),
                  let head = boxedBridge.bridge.interestSetFromEvent(
                      event: gatedEvent.event.bridgeEvent(bridge: boxedBridge.bridge),
                      accountPubkey: account
                  ) as? [String: Any],
                  let hashtags = head["hashtags"] as? [String],
                  let createdAt = (head["createdAt"] as? NSNumber)?.int64Value else { return nil }
            return InterestHead(hashtags: hashtags, createdAt: Int64(createdAt))
        }
    }

    /// Replaceable head: only accept heads newer than the current one.
    private func absorbHead(_ head: InterestHead) {
        guard head.createdAt > headAt else { return }
        headAt = head.createdAt
        hashtags = Set(head.hashtags)
    }
}

File: apps/android/src/main/kotlin/space/bitos/app/data/feed/HashtagFollowsStore.kt
package space.bitos.app.data.feed

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import space.bitos.app.data.relay.RelayPool
import space.bitos.core.model.InterestSet
import space.bitos.core.model.NostrEvent
import space.bitos.core.nostr.EventHasher
import space.bitos.core.nostr.NostrEventCodec
import space.bitos.core.nostr.Sha256EventHasher

/**
 * NIP-51 followed hashtags (kind 30015 d=interest) — web `hashtag-follows`
 * parity: the account's interest-set head syncs from relays (newest verified
 * wins); toggles flip optimistically and publish through the receipt
 * machine. One instance per process.
 */
class HashtagFollowsStore(
    private val scope: CoroutineScope,
    private val pool: RelayPool,
    private val hasher: EventHasher = Sha256EventHasher,
) {
    private val mutableState = MutableStateFlow<Set<String>>(emptySet())
    val hashtags: StateFlow<Set<String>> = mutableState.asStateFlow()

    private var accountPubkey: String? = null
    private var head: NostrEvent? = null
    private var collectJob: Job? = null
    private var requested = false

    init {
        collectJob = scope.launch {
            pool.frames.collect { frame ->
                val account = accountPubkey ?: return@collect
                val event = runCatching {
                    NostrEventCodec.decodeRelayEvent(hasher, frame.message, frame.relay)
                }.getOrNull() ?: return@collect
                if (!NostrEventCodec.verifySignature(hasher, event)) return@collect
                if (event.kind != InterestSet.KIND || event.pubkey.value != account) return@collect
                // Replaceable head: newest verified list wins.
                if (head != null && event.createdAt <= head!!.createdAt) return@collect
                head = event
                mutableState.value = InterestSet.followedHashtags(event).toSet()
            }
        }
        pool.start()
    }

    /** Account lifecycle: re-opens the interest-set head REQ. */
    fun setAccount(pubkey: String?) {
        accountPubkey = pubkey
        head = null
        requested = false
        mutableState.value = emptySet()
        subscribe()
    }

    /** Optimistic local flip; the caller publishes the resulting set. */
    fun toggle(tag: String): Set<String> {
        val normalized = InterestSet.normalize(tag)
        if (!InterestSet.isValid(normalized)) return mutableState.value
        val updated = if (normalized in mutableState.value) {
            mutableState.value - normalized
        } else {
            mutableState.value + normalized
        }
        mutableState.value = updated
        return updated
    }

    fun isFollowed(tag: String): Boolean =
        InterestSet.normalize(tag) in mutableState.value

    private fun subscribe() {
        val account = accountPubkey ?: return
        if (requested) return
        requested = true
        val filter = """{"kinds":[${InterestSet.KIND}],"authors":["$account"],"#d":["${InterestSet.D_TAG}"],"limit":1}"""
        pool.broadcast(NostrEventCodec.encodeRequest("bitos-interest", filter))
    }
}
